### PR TITLE
feature/message_history

### PR DESCRIPTION
**Pull request by PetrochukKsenija**

### CODE REVIEW

File: problem-solver/cxx/dialogControlModule/agent/MessageHistoryGenerationAgent.hpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace dialogControlModule
+{
+class MessageHistoryGenerationAgent : public ScAgent
+{
+  SC_CLASS(Agent, Event(scAgentsCommon::CoreKeynodes::question_initiated, ScEvent::Type::AddOutputEdge))
+  SC_GENERATED_BODY()
+
+private:
+  bool checkActionClass(const ScAddr & actionNode);
```
Comment (MksmOrlov) [inline problem-solver/cxx/dialogControlModule/agent/MessageHistoryGenerationAgent.hpp:19]: According to the codestyle
```suggestion
  bool checkActionClass(ScAddr const & actionNode);
```

File: problem-solver/cxx/dialogControlModule/generator/MessageHistoryGenerator.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  ScAddr lastMessageAddr;
+
+  // if (!dialogAddr.IsValid())
+  //   SC_THROW_EXCEPTION(ExceptionItemNotFound, "Can't add message to dialog. Dialog addr is not valid");
+  // if (!messageAddr.IsValid())
+  //   SC_THROW_EXCEPTION(ExceptionItemNotFound, "Can't add message to dialog. Message addr is not valid");
+  // if (!authorsAddr.IsValid())
+  //   SC_THROW_EXCEPTION(ExceptionItemNotFound, "Can't add message to dialog. Authors addr is not valid");
+  // if (IteratorUtils::getAnyFromSet(context, authorsAddr).IsValid() == SC_FALSE)
+  //   SC_THROW_EXCEPTION(ExceptionItemNotFound, "Can't add message to dialog. Author addr is not valid");
```
Comment (MksmOrlov) [inline problem-solver/cxx/dialogControlModule/generator/MessageHistoryGenerator.cpp:32]: remove commented code

File: problem-solver/cxx/dialogControlModule/keynodes/MessageKeynodes.hpp [hunk 3]
```diff
@@ -36,7 +39,7 @@ class MessageKeynodes : public ScObject
 
   SC_PROPERTY(Keynode("nrel_answer_pattern"), ForceCreate)
   static ScAddr nrel_answer_pattern;
-
+
```
Comment (MksmOrlov) [inline problem-solver/cxx/dialogControlModule/keynodes/MessageKeynodes.hpp:42]: remove this space

File: kb/section_subject_domain_of_dialogues/section_subject_domain_of_dialog_control/examples/message_processing_program.scs
Comment (MksmOrlov) [inline kb/section_subject_domain_of_dialogues/section_subject_domain_of_dialog_control/examples/message_processing_program.scs:1]: You may not change scs program. It is already removed on develop. You can rebase on develop to get actual state

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: docs/agents/messageHistoryGenerationAgent.md
Comment (MksmOrlov) [inline docs/agents/messageHistoryGenerationAgent.md:1]: rename file

File: docs/index.md [hunk 4]
```diff
@@ -14,6 +14,7 @@
     * [Message reply agent](agents/messageReplyAgent.md)
     * [Standard message reply agent](agents/standardMessageReplyAgent.md)
     * [Message topic classification agent](agents/messageTopicClassificationAgent.md)
+    * [Message history generation agent](agents/messageHistoryGenerationAgent.md)
```
Comment (MksmOrlov) [inline docs/index.md:17]: And rename page

File: problem-solver/cxx/dialogControlModule/agent/MessageHistoryGenerationAgent.cpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+#include "sc-agents-common/utils/CommonUtils.hpp"
+#include "sc-agents-common/utils/IteratorUtils.hpp"
+
+#include "MessageHistoryGenerationAgent.hpp"
+
+using namespace utils;
+
+namespace dialogControlModule
+{
+SC_AGENT_IMPLEMENTATION(MessageHistoryGenerationAgent)
```
Comment (MksmOrlov) [inline problem-solver/cxx/dialogControlModule/agent/MessageHistoryGenerationAgent.cpp:16]: Rename the agent and the file according to the action class

File: problem-solver/cxx/dialogControlModule/agent/MessageHistoryGenerationAgent.cpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    SC_LOG_DEBUG("MessageHistoryGenerationAgent: the action doesn't have a message node");
+    SC_LOG_DEBUG("MessageHistoryGenerationAgent finished");
+    AgentUtils::finishAgentWork(&m_memoryCtx, actionNode, false);
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+
+  ScAddr dialogAddr = utils::IteratorUtils::getAnyByOutRelation(&m_memoryCtx, actionNode, scAgentsCommon::CoreKeynodes::rrel_2);
+
+  MessageHistoryGenerator generator(&m_memoryCtx);
+  generator.addMessageToDialog(dialogAddr, message);
```
Comment (MksmOrlov) [inline problem-solver/cxx/dialogControlModule/agent/MessageHistoryGenerationAgent.cpp:39]: Make try catch. You throw exception in the generator method

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
1. Fix old unresolved comments
2. Fix new comments
3. Rebase on develop

File: problem-solver/cxx/dialogControlModule/CMakeLists.txt [hunk 7]
```diff
@@ -37,6 +41,8 @@ include_directories(
     ${SC_COMMON_MODULE_SRC}
 )
 
+
+
```
[the remark this answers is not in the document]
Comment (PetrochukKsenija) [reply problem-solver/cxx/dialogControlModule/CMakeLists.txt]: i had removed this lines

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: kb/section_subject_domain_of_dialogues/section_subject_domain_of_dialog_control/examples/message_processing_program.gwf
Comment (MksmOrlov) [inline kb/section_subject_domain_of_dialogues/section_subject_domain_of_dialog_control/examples/message_processing_program.gwf:1]: After action_find_word_in_set_by_first_letter must be **nrel_goto**, not **nrel_then**. The system doesn't answer. The same as in develop
image

File: kb/test_dialog.scs [hunk 8]
```diff
@@ -0,0 +1,5 @@
+test_dialog
+=> nrel_main_idtf:
+    [тестовый дилог]
```
Comment (MksmOrlov) [inline kb/test_dialog.scs:3]:
```suggestion
    [тестовый диалог]
```

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: problem-solver/cxx/messageReplyModule/generator/MessageHistoryGenerator.cpp [hunk 9]
```diff
@@ -25,7 +26,7 @@ void MessageHistoryGenerator::addMessageToDialog(const ScAddr & dialogAddr, cons
       ScType::EdgeAccessConstPosPerm,
       ScType::NodeConst,
       ScType::EdgeAccessConstPosPerm,
```
Comment (MksmOrlov) [inline problem-solver/cxx/messageReplyModule/generator/MessageHistoryGenerator.cpp:28]:
```suggestion
      ScType::EdgeAccessConstPosTemp,
```
Check all arcs to be temporal
image

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: problem-solver/cxx/messageReplyModule/generator/MessageHistoryGenerator.cpp (hunk 9 above)
Comment (MksmOrlov) [reply problem-solver/cxx/messageReplyModule/generator/MessageHistoryGenerator.cpp:28]: Remove duplicated generator in message reply module

File: problem-solver/cxx/messageReplyModule/test/agent/GenerateReplyMessageAgent.hpp
Comment (MksmOrlov) [inline problem-solver/cxx/messageReplyModule/test/agent/GenerateReplyMessageAgent.hpp:1]: What happened to this file? ^)

File: docs/agents/images/addMessageToTheDialogAgentOutputDialog.png
Comment (MksmOrlov) [inline docs/agents/images/addMessageToTheDialogAgentOutputDialog.png:1]: I can't find constructions like at the image (nrel_dialogue_history, participants, nika).
image
1. Create constructions like at the image. See `interface/src/api/sc/agents/resolveUserAgent.ts`, replace `rrel_dialog_participants` with `nrel_dialog_participants`
2. Rename `nrel_dialog_history` in the image to the `nrel_dialogue_history`

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Review by MksmOrlov — CHANGES_REQUESTED** [review]
- Formalize nrel_dialogue_history
- Rebase on develop and add russian documentation

File: kb/extra/nrel_dialogue_history.scs [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+<- relation;
+<- binary_relation;
+<- oriented_relation;
+<- antireflexive_relation;
+<- asymmetric_relation;
+<- antitransitive_relation;
+=> nrel_main_idtf:
+	[История диалога*] (* <- lang_ru;; *);
+=> nrel_main_idtf:
+	[Dialogue history*] (* <- lang_en;; *);;
```
Comment (MksmOrlov) [inline kb/extra/nrel_dialogue_history.scs:12]: endline

File: docs/agents/addMessageToTheDialogAgent.ru.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
+Данный агент добавляет сообщение в диалога.
```
Comment (MksmOrlov) [inline docs/agents/addMessageToTheDialogAgent.ru.md:1]: в диалог

File: docs/agents/addMessageToTheDialogAgent.ru.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+**Параметры:**
+
+1. `message node` - элемент класса `concept_message` и `concept_atomic_message` или `concept_non_atomic_message`;
+2. `dialog history` - элемент класса `concept_dialogue`, в котором будет сохранена история диалога.
+
+**Ход работы агента:**
+
+* Узел сообщения, полученный агентом, добавляется к последовательности сообщений в истории диалога.
+* Если история диалогов пуста, то к первому добавленному сообщению будет добавлено отношение `rrel_1`. Последнее сообщение всегда имеет отношение `rel_last`. Между сообщениями существует отношение `nrel_message_sequence`, которое определяет последовательность сообщений в диалоге.
```
Comment (MksmOrlov) [inline docs/agents/addMessageToTheDialogAgent.ru.md:15]: Если история диалог**а**

File: docs/agents/addMessageToTheDialogAgent.ru.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+### Пример
+
+Пример структуры ввода для добавления пользовательского сообщения в диалог:
```
Comment (MksmOrlov) [inline docs/agents/addMessageToTheDialogAgent.ru.md:19]: Пример входной структуры. Для добавления сообщения пользователя

File: docs/agents/addMessageToTheDialogAgent.ru.md [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+<img src="../images/addMessageToTheDialogAgentInputUser.png"></img>
+<img src="../images/addMessageToTheDialogAgentInputDialogUser.png"></img>
+
+Пример структуры ввода для добавления сообщения системы в диалог:
```
Comment (MksmOrlov) [inline docs/agents/addMessageToTheDialogAgent.ru.md:24]: Пример входной структуры

File: docs/agents/addMessageToTheDialogAgent.ru.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+<img src="../images/addMessageToTheDialogAgentInputNika.png"></img>
+<img src="../images/addMessageToTheDialogAgentInputDialogNika.png"></img>
+
+Пример полученной структуры:
```
Comment (MksmOrlov) [inline docs/agents/addMessageToTheDialogAgent.ru.md:29]: Пример выходной структуры

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

**Pull request merged**